pypowerwall/tedapi/decorators.py

```diff
@@ -17,3 +19,57 @@ def wrapper(*args, **kwargs):
         return func(*args, **kwargs)
 
     return wrapper
+
+
+# Connection Decorator
+# Checks to see whether a connection exists, and if not, attempts to connect
+def uses_connection_required(func):
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if not self.din:
+            if not self.connect():
+                log.error("Not Connected - Unable to get status")
+                return None
+        return func(self, *args, **kwargs)
+    return wrapper
+
+
+# Cache Decorator
+# Checks to see whether a cached result exists for the function, and if so, returns it. If not, it calls the function and caches the result.
+def uses_cache(cache_key):
+    def decorator(func):
+        def wrapper(self, *args, **kwargs):
+            # We probably can't guarantee that "force" will be the first argument for the function, so let's find it
+            sig = inspect.signature(func)
+            namedArgs = sig.bind(self, *args, **kwargs).arguments
+            force = False
+            if 'force' in namedArgs:
+                force = namedArgs['force']
+
+            # Allow for dynamic cache keys using an argument to the function
+            # e.g. @uses_cache('battery-[args-din]') would cache the result of the function with the key 'battery-<din>'
+            use_cache_key = cache_key
+            if '[args-' in cache_key:
+                namedArg = cache_key.split('args-')[1].split(']')[0]
```

Comment:
Would a regex be better to extract the values we want here?

Reply:
That'd be my preference, but it scares people when I do that lol

Reply:
Use the right tool for the job. If some people protest that it's scary or they can't understand it; that's not YOUR problem! (IE: Healthy boundaries -- It's theirs!)

Split is a great way to do ersatz regex, and if this works it works, but any more complex than this and I'd probably say you'd want a regex.